apis/Google.Cloud.Firestore.Data/Google.Cloud.Firestore.Data/DocumentSnapshot.cs

```diff
@@ -0,0 +1,211 @@
+using Google.Api.Gax;
+using Google.Cloud.Firestore.V1Beta1;
+using System;
+using System.Collections.Generic;
+using System.Dynamic;
+
+namespace Google.Cloud.Firestore.Data
+{
+    /// <summary>
+    /// An immutable snapshot of the data for a document.
+    /// </summary>
+    public sealed class DocumentSnapshot
+    {
+        /// <summary>
+        /// The proto representation of the document. Primarily visible for testing purposes.
+        /// </summary>
+        internal Document Document { get; }
+
+        /// <summary>
+        /// The full reference to the document.
+        /// </summary>
+        public DocumentReference Reference { get; }
+
+        /// <summary>
+        /// The database that owns the document.
+        /// </summary>
+        public FirestoreDb Database => Reference.Database;
+
+        /// <summary>
+        /// The ID of the document.
+        /// </summary>
+        public string Id => Reference.Id;
+
+        /// <summary>
+        /// Whether or not the document exists.
+        /// </summary>
+        public bool Exists => Document != null;
+
+        /// <summary>
+        /// The creation time of the document if it exists, or null otherwise.
+        /// </summary>
+        public Timestamp? CreateTime { get; }
+
+        /// <summary>
+        /// The update time of the document if it exists, or null otherwise.
+        /// </summary>
+        public Timestamp? UpdateTime { get; }
+
+        /// <summary>
+        /// The time at which this snapshot was read.
+        /// </summary>
+        public Timestamp ReadTime { get; }
+
+        private DocumentSnapshot(FirestoreDb db, string resourceName, Document document, Timestamp readTime)
+        {
+            GaxPreconditions.CheckNotNull(db, nameof(db));
+            GaxPreconditions.CheckNotNull(resourceName, nameof(resourceName));
+            Reference = db.GetDocumentReferenceFromResourceName(resourceName);
+            Document = document;
+            CreateTime = Timestamp.FromProtoOrNull(document?.CreateTime);
+            UpdateTime = Timestamp.FromProtoOrNull(document?.UpdateTime);
+            ReadTime = readTime;
+        }
+
+        internal static DocumentSnapshot ForDocument(FirestoreDb db, Document document, Timestamp readTime) =>
+            new DocumentSnapshot(db, document.Name, document, readTime);
+
+        internal static DocumentSnapshot ForMissingDocument(FirestoreDb db, string name, Timestamp readTime) =>
+            new DocumentSnapshot(db, name, null, readTime);
+
+        /// <summary>
+        /// Returns the document data as an <see cref="ExpandoObject"/>, for purposes of dynamic typing.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">This snapshot represents a missing document, i.e. <see cref="Exists"/> is false.</exception>
+        /// <returns>An <see cref="ExpandoObject"/> containing the document data.</returns>
+        public dynamic ToExpando()
```

Comment:
This wasn't suggested by the API team at all - it was purely my own idea, thinking it might be handy (in the same way that I think Json.NET has a dynamic approach available). If we decide we don't want this, ValueDeserializer can be simpler. I'll remove it for now, but leave the infrastructure in place in ValueDeserializer so we can revisit it easily.